// @yemreak/yt-dlp/dist/types.js
export var Acodec;
(function (Acodec) {
    Acodec["Mp4A402"] = "mp4a.40.2";
    Acodec["Mp4A405"] = "mp4a.40.5";
    Acodec["None"] = "none";
    Acodec["Opus"] = "opus";
})(Acodec || (Acodec = {}));
export var SubtitleExt;
(function (SubtitleExt) {
    SubtitleExt["Json3"] = "json3";
    SubtitleExt["Srv1"] = "srv1";
    SubtitleExt["Srv2"] = "srv2";
    SubtitleExt["Srv3"] = "srv3";
    SubtitleExt["Ttml"] = "ttml";
    SubtitleExt["Vtt"] = "vtt";
})(SubtitleExt || (SubtitleExt = {}));
export var Protocol;
(function (Protocol) {
    Protocol["HTTPS"] = "https";
    Protocol["M3U8Native"] = "m3u8_native";
    Protocol["Mhtml"] = "mhtml";
})(Protocol || (Protocol = {}));
export var DynamicRange;
(function (DynamicRange) {
    DynamicRange["SDR"] = "SDR";
})(DynamicRange || (DynamicRange = {}));
export var AudioEXTEnum;
(function (AudioEXTEnum) {
    AudioEXTEnum["M4A"] = "m4a";
    AudioEXTEnum["Mhtml"] = "mhtml";
    AudioEXTEnum["Mp4"] = "mp4";
    AudioEXTEnum["None"] = "none";
    AudioEXTEnum["Webm"] = "webm";
})(AudioEXTEnum || (AudioEXTEnum = {}));
export var Container;
(function (Container) {
    Container["M4ADash"] = "m4a_dash";
    Container["Mp4Dash"] = "mp4_dash";
    Container["WebmDash"] = "webm_dash";
})(Container || (Container = {}));
export var Accept;
(function (Accept) {
    Accept["TextHTMLApplicationXHTMLXMLApplicationXMLQ09Q08"] = "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8";
})(Accept || (Accept = {}));
export var AcceptLanguage;
(function (AcceptLanguage) {
    AcceptLanguage["EnUsEnQ05"] = "en-us,en;q=0.5";
})(AcceptLanguage || (AcceptLanguage = {}));
export var SECFetchMode;
(function (SECFetchMode) {
    SECFetchMode["Navigate"] = "navigate";
})(SECFetchMode || (SECFetchMode = {}));
